examples/demo-rollup/stf/src/hooks_impl.rs

```diff
@@ -24,9 +27,27 @@ impl<C: Context, Da: DaSpec> TxHooks for Runtime<C, Da> {
         arg: RuntimeTxHook<C>,
     ) -> anyhow::Result<C> {
         let RuntimeTxHook { height, sequencer } = arg;
+
+        // fetch the sender and sequencer addresses
         let AccountsTxHook { sender, sequencer } =
             self.accounts
-                .pre_dispatch_tx_hook(tx, working_set, sequencer)?;
+                .pre_dispatch_tx_hook(tx, working_set, sequencer.clone())?;
+
+        // transfer the initial gas from the sender
+        self.bank.pre_dispatch_tx_hook(
+            tx,
+            working_set,
+            BankTxHook {
+                sender: sender.clone(),
+                sequencer: sequencer.clone(),
+            },
+        )?;
+
+        let sender =
+            Arc::try_unwrap(sender).map_err(|e| anyhow::anyhow!("synchronization failure: {e}"))?;
```

Comment:
It would be better if `Context` also keeps `Arc` as well so there will be no need for `try_unwrap` here

Reply:
I moved the `Arc` to `Rc` as we don't perform the hooks under `async`. Is it ok if we move this change to a dedicated PR? Any changes to the `Context` will imply a lot of new LOC to this one

Reply:
One easy way to achieve that is to make the constructor of the context receive anything that can be an arc:

```rust
/// Constructor for the Context.
fn new<S, Q>(sender: S, sequencer: Q, height: u64) -> Self
where
    S: Into<Arc<Self::Address>>,
    Q: Into<Arc<Self::Address>>;
}

Wdyt?

Reply:
Regarding the problem above, I was able to get away with that since `Arc/Rc` employs a hack that returns the rc itself instead of the error.

This allows us to clone the value for the worst case scenario - that is if the value still lives elsewhere. To be fair, this would happen only if the module is trying to do something very weird, such as sending the rc into a bounded channel, but then we have to shield against this possible outcome.

Reply:
Do we need the `Rc` at all? What is the original reason for introducing it?

Reply:
The reason is to avoid unnecessary cloning here

https://github.com/Sovereign-Labs/sovereign-sdk/blob/8472d1d76dc8bb25dcc98ae88e0e87ed7c08c655/examples/demo-rollup/stf/src/hooks_impl.rs#L23-L47

While avoiding lifetime complexity when declaring the types. We can pass them around as owned without bounding lifetimes to the runner and/or context

Reply:
(we do clone there, but we clone the `RC` and not the data itself)

Reply:
Ok, but why is the ownership/cloning needed? I understand we `Rc::clone` cheaply, but why is there a reason for cloning? Even with `Rc(sender )`, the `pre_dispatch_tx_hook` doesn't have access to the owned sender if it doesn't `Rc::unwrap` in the module code.

Reply:
We pass the ownership to `BankTxHook`. The alternative to that would be creating `BankTxHook<'a>`, that introduces lifetime complexity. Its easier to read the code without lifetime mess.

If we really want the lifetime there, we would have to introduce a lifetime to the bank hook declaration too.

Reply:
You mean changing the signature of the trait like this?

```diff
 /// Runs just before a transaction is dispatched to an appropriate module.
 fn pre_dispatch_tx_hook(
     &self,
     tx: &Transaction<Self::Context>,
     working_set: &mut WorkingSet<Self::Context>,
-    arg: Self::PreArg,
+    arg: &Self::PreArg,
 ) -> anyhow::Result<Self::PreResult>;
```

That makes sense, yes

Reply:
Why wouldn't it work if we do the following:
1. Create `BankTxHook` ` with owned values.
2. Pass  `&BankTxHook` to the `bank.pre_dispatch_tx_hook` so we only borrow.
3. We use `sender & sequencer` to construct the Context, they are still in the original `BankTxHook` which was only borrowed and the borrow ended so now can continue to use it?

Reply:
We must use either owned everywhere or ref everywhere as we share the trait signature. There is an alternative to using Rc if we consume the type on bank and return the argument. I added it on https://github.com/Sovereign-Labs/sovereign-sdk/pull/1223/commits/4ab9d782c534a8b0f8a6adf78bee0c40efed73c0 ; let me know wdyt

Reply:
Fixed on https://github.com/Sovereign-Labs/sovereign-sdk/pull/1223/commits/1ee482aebef655beb365669c00b018318f163129